### PR TITLE
Issue #3173: moved SingleSpaceSeparatorCheck from sevntu to checkstyle

### PR DESCRIPTION
**Pull request by rnveach**
Issue #3173

Start of new check from sevntu.

Made some minor modifications from sevntu.
* Removed references of `nodes` and replaced them with `whitespace` characters.
* Changed property `validateCommentNodes` to `validateComments`.
* Changed check's method definitions from `(DetailAST ast, String line)` to `(String line, int columnNo)`.

Questions:
(1)
````
+    public int[] getDefaultTokens() {
+        return CommonUtils.EMPTY_INT_ARRAY;
+    }
....
+    public void beginTree(DetailAST rootAST) {
+        visitEachToken(rootAST);
+    }
````

Can we just change `getDefaultTokens` to all tokens? I dislike parsing the entire tree twice (once by `TreeWalker` and then again by this check).
Should we use reflection to get all tokens, or should I hardcode the list and make a test to verify?

(2)
````
+    private static boolean isSpace(String line, int columnNo) {
+        return line.charAt(columnNo) == ' ';
+    }
````

Is there a specific reason we don't just use `isWhitespace`? is there a reason we don't check tabs and just a space?

### CODE REVIEW

**Comment by rnveach** [comment]
Travis failure is because I forgot Issue reference. It has been fixed now.

**Comment by romani** [comment]
1) extremely weird why we have no exception as you mention in config old name "validateCommentNodes" , but there is no such property in Check.

2)
> Can we just change getDefaultTokens to all tokens? I dislike parsing the entire tree twice (once by TreeWalker and then again by this check).

It is better to do traversing separately in this Check, rather than be called on each node. Logic of Check is very simple so it is ok.

3)
I do not remember but most likely due to focus on this Check on spaces (not a "\t", "\n", ...)

**Comment by rnveach** [comment]
> extremely weird why we have no exception

Good catch.
The reason was it was listed as `metadata` and not `property`.
If we should throw exceptions for unknown properties, I can make an issue for it. I don't know the reason for metadata tags.

**Comment by romani** [comment]
> I don't know the reason for metadata tags.

https://github.com/checkstyle/checkstyle/issues/2726 we already had this issue.

**Comment by rnveach** [comment]
Fixed 1 and 4. Left 2 and 3 as is.

**Comment by romani** [comment]
appveyor failure - "[ERROR] src\main\java\com\puppycrawl\tools\checkstyle\checks\whitespace\SingleSpaceSeparatorCheck.java:[110] (extension) SimpleAccessorNameNotation: Unexpected getter name. "

**Comment by rnveach** [comment]
Fixed.

**Comment by romani** [comment]
code is good and ready to merge.
@rnveach , thanks a lot !

I do not remember that we did stress testing of this Check by checkstyle-tester.

We need to update checkstyle-tester config to have this Check.

@rnveach , please provide checkstyle-tester report:
- for exception testing against all projects
- for functional testing against some project (make a decision yourself base on number or violations)

File: src/main/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/messages_de.properties [hunk 1]
```diff
@@ -18,3 +18,5 @@ ws.illegalFollow=Nach ''{0}'' folgt ein ungültiges Zeichen.
 ws.typeCast=Nach ''typumwandlung'' folgt kein Leerzeichen.
 
 empty.line.separator.multiple.lines = ''{0}'' verfügt über mehr als 1 Leerzeilen vor.
+
+single.space.separator=Verwenden Sie einen einzigen Raum ohne Leerzeichen zu trennen.
```
Comment (robertpainsi) [inline src/main/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/messages_de.properties]: translates to `Use a single room without space to separate.`. :joy_cat:

If it's useful for this PR, use something like `Verwenden Sie ein einfaches Leerzeichen um Zeichen zu trennen.`
[Translation from translate.google.com](https://translate.google.com/#de/en/Verwenden%20Sie%20ein%20einfaches%20Leerzeichen%20um%20Zeichen%20zu%20trennen)

Comment (rnveach) [inline src/main/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/messages_de.properties]: @robertpainsi Thanks, I will update it.

**Comment by rnveach** [comment]
Made the changed request by @robertpainsi.

@romani 
Here is the whole report. http://rveach.no-ip.org/checkstyle/regression/reports/19/index.html
No exceptions. Orekit has the most violations, 8599.
jOOL has a managable 23.

**Comment by rnveach** [comment]
When running with comments on, I did get one exception.
````
Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing /home/ricky/opensource/contribution/checkstyle-tester/src/main/java/checkstyle-checkstyle-f9ce38f/src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputNewlineCrAtEndOfFile.java
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: 130
    at java.lang.String.charAt(String.java:658)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.SingleSpaceSeparatorCheck.isPrecededByMultipleWhitespaces(SingleSpaceSeparatorCheck.java:204)
````

I can't reproduce the exception on Windows. Since the file is about `\r` and inside the comments, it could be related to #3066.

**Comment by romani** [comment]
> Orekit has the most violations, 8599

yes, this Check is not for this project. Orekit has different style.

> When running with comments on, I did get one exception

At what project/file you got this Exception ?
If it is from our our checkstyle-tester list , we can not merge this Check for now 
Please send PR for https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/checks-nonjavadoc-error.xml to have a new Check in it.

**Comment by robertpainsi** [comment]
@\2
This problem has been mentioned at https://groups.google.com/forum/#!msg/sevntu-checkstyle/kSCClhhf_hs/dIZJjodyCAAJ point 4. I agree with @rnveach that https://github.com/checkstyle/checkstyle/issues/3066 is related and may also solve it.

@\3
On each reported line, there are two spaces used as separator.
At [MapStream.html#L131](http://rveach.no-ip.org/checkstyle/regression/reports/19/protonpack/xref/poetix-protonpack-36d5801/src/main/java/com/codepoetics/protonpack/maps/MapStream.html#L131) there are two spaces after `value3,`.
In [ZipTest.html](http://rveach.no-ip.org/checkstyle/regression/reports/19/protonpack/xref/poetix-protonpack-36d5801/src/test/java/com/codepoetics/protonpack/ZipTest.html) line 17, 27, 37, 60 there are two spaces before the equals sign.

**Comment by rnveach** [comment]
> At what project/file you got this Exception ?

This is from our own project, Checkstyle, with the file `InputNewlineCrAtEndOfFile.java`.

> we can not merge this Check for now 

Until we fix the root error? (references issue)

> it is nightmare to find a extra space - we need to print column number in message.

We do print column number as seen in our junits.
Since this is a new check and we are not using `patch-diff-report-tool`, I assume this is the limitation of the fallback report. The `checkstyle-result.xml` file has the column numbers listed.

> why there is mismatch ?

Was a previous issue with me running regression. I will rerun the reports.

**Comment by rnveach** [comment]
@romani Here is the fixed regression.
http://rveach.no-ip.org/checkstyle/regression/reports/34/

**Comment by romani** [comment]
> Until we fix the root error?

kind of yes. In other case there will CI failure.
Or we could add one more exclude, for this file only, to https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/projects-for-travis.properties#L6

**Pull request merged**